### DISCUSSION TITLE
KDL data model

### DISCUSSION THREAD

**Discussion by nichtich** [kdl-org/kdl#183] [Ideas]
KDL specification does not explicitly define a data model but only a syntax that implies some data model of KDL. As long as you parse from KDL syntax into some data structure of a programming language, a KDL data model is not needed. In contrast both KQL and KDL Schema build on a data model. The lack of its explicit definition leaves open handling of some edge cases, such as:

* Is an empty string tag different to no tag ( does `("")value` differ from `value`)?
* Is an empty list of child nodes different from no child nodes (does `node {}` differ from `node`)?

The data model should abstract from syntax details such as comments, how nodes are terminated, whether names are given as bare identifiers or given as strings etc.

Here is a draft of a possible data model (without explicit definition of *name* and *value* and both questions above answered with no):

* a *document* is a list of *nodes*
* a *node* consist of
  * a *tag* being a *name* 
  * a *name*
  * a list of *arguments*, each being a *value*
  * a set of *properties*, each consisting of a *name* and *value*, and each *name* being unique
  * a list of *children* being a *document*

**Comment by larsgw** [comment]
Looks good IMO.

> * Is an empty string tag different to no tag ( does `("")value` differ from `value`)?

Yes, I think similarly node names can also be just `""` (so the draft does not change). See also https://github.com/kdl-org/kdl/discussions/177.

> * Is an empty list of child nodes different from no child nodes (does `node {}` differ from `node`)?

Not sure, but I do know that kdljs does not distinguish between the two but kdl4j does.

**Reply by hkolbeck** [reply]
I made the distinction in kdl4j since it seemed odd for someone to specify a present-but-empty child, and so they might mean something specific by it. I am not attached to that choice.

**Reply by nichtich** [reply]
Whether `("")value` differs from `value` and whether `node` differs from `node {}` is a matter of choice of the KDL parser, so it is not specified yet. This is not important for several use cases but at least KDL Query Language should be clear about what to expect.

**Reply by tabatkins** [reply]
Currently the tests are clear that `("")value` and `value` are different, and `node` and `node {}` are different.

**Reply by zkat** [reply]
and that seems fine to be tbh

**Reply by nichtich** [reply]
Ok if both child nodes and tags are nullable, the abstract KDL data model should be:

* a *document* is a list of *nodes*
* a *node* consist of
  * a *name*
  * an optional *tag* being a *name* 
  * a list of *arguments*, each being a *value*
  * a set of *properties*, each consisting of a *name* and *value*, and each *name* being unique in the set
  * an optional list of *children* being a *document*
* a *name* is a list of Unicode code points
* a *value* is either 
  * a list of Unicode code points
  * a *number*, consisting of an integer part and a decimal part, 
  * a *boolean*, being one of the special values *true* and *false*, 
  * or the special value *null*

KDL defines a syntax of this data model but KQL and KDL purely act on the abstract KDL data structures independently from syntax.
